Cannot authenticate to Github repo during the installation process

<!--
**Please only file bugs/feature requests for pnpm here.**

If you have a question that is not answered in our FAQ: https://pnpm.js.org/en/faq
then it is best to ask it in our Discord chatroom:
https://discord.gg/mThkzAT

If your issue is a bug, please follow the format below:
-->

When trying to install my packages using a private registry, `pnpm` goes in error.
I think I know what's going on.

Doing the same via `npm` works fine because it installs the dependencies that are not in scope from the npm registry and installs the scoped dependencies (the private ones) from the github registry. I could verifying it by checking its `package-lock.json`

On the other hand, `pnpm` tries to install ALL the dependencies from one single registry, which is github.

### pnpm version: 5.9.3

### Code to reproduce the issue:

Code inside `.npmrc`

```
registry=https://npm.pkg.github.com/{SCOPE}
```

I can't give the information of the log because it contains sensitive information.


### Expected behavior:

It should install all the packages.

### Actual behavior:

```
 ERROR  GET https://npm.pkg.github.com/{SCOPE}/bootstrap: Unauthorized - 401
No authorization header was set for the request.

These authorization settings were found:
//npm.pkg.github.com/:_authToken={TOKEN}
```


### Additional information:

Changing `.npmrc` to the following works fine

```
{SCOPE}:registry=https://npm.pkg.github.com/{SCOPE}
```

Any progress on this please? Haven't been able to use `pnpm` because of this for a long time. My situation:
`.npmrc`
```
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=<token>
```
`pnpm i`
```
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/apify-packages/@apify-packages%2Ffingerprint-injector: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=72d4[hidden]
Progress: resolved 1, reused 0, downloaded 0, added 0
```
Any advice would be highly appreciated.

I have the identical issue, this does seem like a long time.
~/.npmrc
```
//npm.pkg.github.com/:_authToken=ghp_XXXXX
//npmjs.org=true
@1uphealth:registry=https://npm.pkg.github.com/1uphealth
```

then...
```
pnpm add @1uphealth/core-1upenv-script
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/1uphealth/@1uphealth%2Fcore-1upenv-script: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
//npm.pkg.github.com/:_authToken=ghp_[hidden]
@1uphealth:registry=https://npm.pkg.github.com/1uphealth
```

Have the same issue:
```
ERR_PNPM_FETCH_401  GET https://<...>/repository/npm-group/@babel/core/-/core-7.16.10.tgz: Unauthorized - 401
No authorization header was set for the request.
These authorization settings were found:
//nexus.redmadrobot.com/repository/npm-private:_authToken=NpmT[hidden]
always-auth=true
```

any news on this? Using pnpm 6.27.1

Hello same issues there with `pnpm 6.25.1`

```
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/@askia%2Fa-b: Unauthorized - 401   

An authorization header was used: Bearer ghp_[hidden]

These authorization settings were found:
@askia:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=ghp_[hidden]
```

We're doing something like this 

```
@org:registry=https://npm.pkg.github.com/org-packages
//npm.pkg.github.com/:_authToken=<token>
```

However we're getting a 404 error: 

```
 ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@flixed-io%2Fatoms: Not Found - 404

@org/package is not in the npm registry, or you have no permission to fetch it.
```

Any idea why we're seeing this?

We're also able to fetch it fine in another repo that's just using `npm` instead of `pnpm`.

I have the following `.npmrc`:

```
registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
```

And til `pnpm@7.13` it was working just fine.

I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:

```
 ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
```

Downgrading to `7.13` worked well

> I have the following `.npmrc`:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> And til `pnpm@7.13` it was working just fine.
> 
> I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> 
> ```
>  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> 
> No authorization header was set for the request.
> 
> These authorization settings were found:
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> ```
> 
> Downgrading to `7.13` worked well

This flip-flopping between versions has been really painful. I haven't got the bandwidth to dig in and find out why atm. Were you able to get that working in CI/CD as well? Just out of curiosity. Thanks so much for sharing

> I have the following `.npmrc`:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> And til `pnpm@7.13` it was working just fine.
> 
> I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> 
> ```
>  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> 
> No authorization header was set for the request.
> 
> These authorization settings were found:
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> ```
> 
> Downgrading to `7.13` worked well

I think this will work for you:
```
registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
//company-artifactory-url.com:443/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
```

Duplicating the `_authToken` entry but including `:443` in the second one. Keep both.

I met  ERR_PNPM_FETCH_404  GET https://company-url/npm/npm-registry/xxxxxx: Not Found - 404

xxxxxx is not in the npm registry, or you have no permission to fetch it.

An authorization header was used: Bearer efd5[hidden]

I downgrade to 7.0 and still got error. but npm is ok.

getting same issue with 6.23.1 and 7.18.1

+ pnpm install --loglevel=verbose
 ERR_PNPM_FETCH_401  GET <Default registry>/@babel%2Fpreset-react: Unauthorized - 401

This error happened while installing a direct dependency of /opt/app/m61407_node/zlp26614.vci.att.com_27531_websocket/workspace/i-template_devdeploy_helmChanges/repository

No authorization header was set for the request.

Also having this issue on 7.22.0 in combination with gitlab. 
Rolled back to 7.13.0 which is working fine.

 ERR_PNPM_FETCH_404  GET <repo-url> Not Found - 404
No authorization header was set for the request.

> > I have the following `.npmrc`:
> > ```
> > registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> > //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And til `pnpm@7.13` it was working just fine.
> > I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> > ```
> >  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> > 
> > No authorization header was set for the request.
> > 
> > These authorization settings were found:
> > //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Downgrading to `7.13` worked well
> 
> I think this will work for you:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> //company-artifactory-url.com:443/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> Duplicating the `_authToken` entry but including `:443` in the second one. Keep both.

I found that duplicating `_authToken_` line with port `:443` worked for me, but I think its perhaps a separate issue.

I got into to the situation like this:

1. Init new project with pnpm 
2. Install some packages from public npm regisry
3. Notice you now need stuff from private npm registry
4. Setup everything for that
5. Now try to install private package
6. It wont work. The tarball has `:443` in the URL so I suspect that could have something to do with the error
7. Add the extra `_authToken` line to your `.npmrc` config.
8. Now install the private package.
9. Package is installed succesfully

I want to add one extra scenario I encountered that does not make any sense to me - the URL is case-sensitive...

Config:
```ini
registry=http://xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/
always-auth=true
shamefully-hoist=true
strict-ssl=false

; begin auth token
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:username=${WWW_NPM_USER}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:_password=${WWW_NPM_TOKEN}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:email=npm requires email to be set but doesn't use the value
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:username=${WWW_NPM_USER}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:_password=${WWW_NPM_TOKEN}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:email=npm requires email to be set but doesn't use the value
; end auth token
```

Issue:
```ini
Packages: +851
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 ERR_PNPM_FETCH_401  GET http://xxx:8080/tfs/Www_BU_Mmm/_packaging/npmProductionFeed/npm/registry/@babel/core/-/core-7.20.12.tgz: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
```

We got private package resolving issues too and the last valid version is `7.13.6`.

Some context: we use a Gitlab instance on premise and our .npmrc look like:

```
//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN
@team-software:registry=https://gitlab.domain.com/api/v4/packages/npm/
```

We try to install a library hosted on this gitlab registry: `"@team-software/v-components": "0.0.96"`

This configuration works with `pnpm<=7.13.6` or `npm`  but with pnpm `8.6.6` (and node `20.3.0`) we get:

```
 WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 10 seconds. 2 retries left.
 WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 1 minute. 1 retries left.
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /builds/.pnpm-store/v3
  Virtual store is at:             node_modules/.pnpm
 ERR_PNPM_FETCH_404  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz: Not Found - 404
 No authorization header was set for the request.
```

I did some tests:

- I added `gitlab.domain.com/api/v4/projects/:_authToken=AUTH_TOKEN` and this fixed the issue
- Then I removed `//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN` this brings a new resolving issue. I guess no match between `@team-software:registry=xxx` and the `xxx:_authToken=token`
- I merged both line and used  `//gitlab.domain.com/api/v4/:_authToken=AUTH_TOKEN`,  **this is working** (but doesn't follow what gitlab recommend to use for npm).

Final:

I feel like to resolve the package there is a first match on `gitlab.domain.com/api/v4/packages/npm/` then a download issued on the url result  `gitlab.domain.com/api/v4/projects/`. Before the header `AUTH_TOKEN` issued on the first match was forwarded on the "related download" even if there was a matching mismatch?

_Just a guess without diving too much how things are working, if someone has more context about this 🙏_

So, how does one install (not even trying publishing) in github actions now?

I'm on pnpm@8.7.0, and trying to use https://gist.github.com/belgattitude/838b2eba30c324f1f0033a797bab2e31, with added `env`:
```yml
    - name: Install dependencies
      shell: bash
      working-directory: ${{ inputs.cwd }}
      run: pnpm install --frozen-lockfile --prefer-offline
      env:
        NPM_TOKEN: ${{ github.token }}
        NPM_AUTH_TOKEN: ${{ github.token }}
        NODE_AUTH_TOKEN: ${{ github.token }}
        npm_config__authtoken: ${{ github.token }}
        npm_config_//npm.pkg.github.com:_authtoken: ${{ github.token }}
```
.npmrc:
```
resolution-mode=highest
strict-peer-dependencies=true
auto-install-peers=false
@my-cool-repo:registry=https://npm.pkg.github.com
```

I get:
```
ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/download/...: Unauthorized - 401
No authorization header was set for the request.
These authorization settings were found:
@lifekeys:registry=https://npm.pkg.github.com/
_authToken=ghs_[hidden]
```

I cannot add `//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}` because then I get `WARN Issue while reading  "/home/..../.npmrc". Failed to replace env in config: ${GITHUB_TOKEN}` when installing locally, and that makes pnpm ignore `auto-install-peers` setting there.

I cannot downgrade pnpm to `7.13.6` in the `action.yml` because option to use lockfile v6 was added in `7.24.0`

what do I do?

> We got private package resolving issues too and the last valid version is `7.13.6`.
> 
> Some context: we use a Gitlab instance on premise and our .npmrc look like:
> 
> ```
> //gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN
> @team-software:registry=https://gitlab.domain.com/api/v4/packages/npm/
> ```
> 
> We try to install a library hosted on this gitlab registry: `"@team-software/v-components": "0.0.96"`
> 
> This configuration works with `pnpm<=7.13.6` or `npm` but with pnpm `8.6.6` (and node `20.3.0`) we get:
> 
> ```
>  WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 10 seconds. 2 retries left.
>  WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 1 minute. 1 retries left.
> Packages are hard linked from the content-addressable store to the virtual store.
>   Content-addressable store is at: /builds/.pnpm-store/v3
>   Virtual store is at:             node_modules/.pnpm
>  ERR_PNPM_FETCH_404  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz: Not Found - 404
>  No authorization header was set for the request.
> ```
> 
> I did some tests:
> 
> * I added `//gitlab.domain.com/api/v4/projects/:_authToken=AUTH_TOKEN` and this fixed the issue
> * Then I removed `//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN` this brings a new resolving issue. I guess no match between `@team-software:registry=xxx` and the `xxx:_authToken=token`
> * I merged both line and used  `//gitlab.domain.com/api/v4/:_authToken=AUTH_TOKEN`,  **this is working** (but doesn't follow what gitlab recommend to use for npm).
> 
> Final:
> 
> My understanding is to resolve the package there is a first match on `gitlab.domain.com/api/v4/packages/npm/` then a download issued on the url result `gitlab.domain.com/api/v4/projects/`. Before the header `AUTH_TOKEN` issued on the first match was forwarded on the "related download" even if there was an uri mismatch?
> 
> _Just a guess without diving too much how things are working, if someone has more context about this 🙏_

You saved me. I spent 3 hours to find this answer and it really works.

Have anyone had similar issues trying to access a Azure DevOps artifact feed? 

I am getting issues similar to this:

```
 ERR_PNPM_FETCH_401  GET https://pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/@<org-alias><project>: Unauthorized - 401

An authorization header was used: Bearer ejVh[hidden]

These authorization settings were found:
//pkgs.dev.azure.com/<org>/_packaging/<project>/npm/:_authToken=ejVh[hidden]
//pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/:_authToken=ejVh[hidden]
@<org-alias>:registry=https://pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/
```

where my local `.npmrc` has the authentication settings shown above.

--- a/.changeset/three-penguins-trade.md
+++ b/.changeset/three-penguins-trade.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/network.auth-header": patch
+"pnpm": patch
+---
+
+Fix missing auth tokens in registries with pathnames specified (e.g. //npm.pkg.github.com/pnpm). #5970 #2933

--- a/.changeset/three-penguins-trade.md
+++ b/.changeset/three-penguins-trade.md
@@ -3,4 +3,4 @@
 "pnpm": patch
 ---
 
-Fix missing auth tokens in registries with pathnames specified (e.g. //npm.pkg.github.com/pnpm). #5970 #2933
+Fix missing auth tokens in registries with paths specified (e.g. //npm.pkg.github.com/pnpm). #5970 #2933

--- a/network/auth-header/src/index.ts
+++ b/network/auth-header/src/index.ts
@@ -24,18 +24,11 @@ function getMaxParts (uris: string[]) {
 }
 
 function getAuthHeaderByURI (authHeaders: Record<string, string>, maxParts: number, uri: string): string | undefined {
-  const url = new URL(uri)
-  const domain = url.host
-
-  // Add a trailing slash to the path, so that we can match it with the auth headers structure
-  const path = url.pathname + '/'
-
-  const fullPathKey = `//${domain}${path}`
-  if (authHeaders[fullPathKey]) {
-    return authHeaders[fullPathKey]
+  if (!uri.endsWith('/')) {
+    uri += '/'
   }
-
   const nerfed = nerfDart(uri)
+  console.log(uri, nerfed)
   const parts = nerfed.split('/')
   for (let i = Math.min(parts.length, maxParts) - 1; i >= 3; i--) {
     const key = `${parts.slice(0, i).join('/')}/`

--- a/network/auth-header/src/index.ts
+++ b/network/auth-header/src/index.ts
@@ -24,6 +24,17 @@ function getMaxParts (uris: string[]) {
 }
 
 function getAuthHeaderByURI (authHeaders: Record<string, string>, maxParts: number, uri: string): string | undefined {
+  const url = new URL(uri)
+  const domain = url.host
+
+  // Add a trailing slash to the path, so that we can match it with the auth headers structure
+  const path = url.pathname + '/'
+
+  const fullPathKey = `//${domain}${path}`
+  if (authHeaders[fullPathKey]) {
+    return authHeaders[fullPathKey]
+  }
+
   const nerfed = nerfDart(uri)
   const parts = nerfed.split('/')
   for (let i = Math.min(parts.length, maxParts) - 1; i >= 3; i--) {

--- a/network/auth-header/src/index.ts
+++ b/network/auth-header/src/index.ts
@@ -28,7 +28,6 @@ function getAuthHeaderByURI (authHeaders: Record<string, string>, maxParts: numb
     uri += '/'
   }
   const nerfed = nerfDart(uri)
-  console.log(uri, nerfed)
   const parts = nerfed.split('/')
   for (let i = Math.min(parts.length, maxParts) - 1; i >= 3; i--) {
     const key = `${parts.slice(0, i).join('/')}/`

--- a/network/auth-header/test/getAuthHeaderByURI.ts
+++ b/network/auth-header/test/getAuthHeaderByURI.ts
@@ -43,3 +43,21 @@ test('getAuthHeaderByURI() when default ports are specified', () => {
 test('returns undefined when the auth header is not found', () => {
   expect(createGetAuthHeaderByURI({ allSettings: {}, userSettings: {} })('http://reg.com')).toBe(undefined)
 })
+
+test('getAuthHeaderByURI() when the registry has pathnames', () => {
+  const getAuthHeaderByURI = createGetAuthHeaderByURI({
+    allSettings: {
+      '//npm.pkg.github.com/pnpm/:_authToken': 'abc123',
+    },
+    userSettings: {},
+  })
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+  expect(getAuthHeaderByURI('https://npm.pkg.github.com/pnpm/foo/-/foo-1.0.0.tgz')).toBe('Bearer abc123')
+})

